ouroboros-consensus/ouroboros-consensus-test-infra/src/Test/Util/HardFork/OracularClock.hs

```diff
@@ -0,0 +1,242 @@
+{-# LANGUAGE Rank2Types          #-}
+{-# LANGUAGE RecordWildCards     #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+-- | Intended for qualified import
+--
+-- > import Test.Util.OracularClock (OracularClock(..))
+-- > import qualified Test.Util.OracularClock as OracularClock
+module Test.Util.HardFork.OracularClock (
+    OracularClock(..)
+  , new
+  , onSlotChange
+  , withinEachSlot
+  , withinTheCurrentSlot
+  ) where
+
+import           Control.Monad (replicateM_, void)
+import           Data.Foldable (toList)
+import           Data.Time
+import           GHC.Stack
+
+import           Ouroboros.Consensus.Util.IOLike
+import           Ouroboros.Consensus.Util.ResourceRegistry
+
+import           Cardano.Slotting.Slot
+
+import qualified Ouroboros.Consensus.BlockchainTime as BTime
+import           Ouroboros.Consensus.Util.STM (onEachChange)
+import           Ouroboros.Consensus.Util.Time (nominalDelay)
+
+import           Test.Util.HardFork.Future (Future, futureSlotLengths,
+                     futureTimeToSlot)
+import           Test.Util.Slots (NumSlots (..))
+import           Test.Util.Stream
+
+-- | A clock that knows the future
+--
+-- This clock's closure contains a 'BTime.SystemTime', a 'Future', and a
+-- 'NumSlots'. As the system time advances, the clock maintains a notion of the
+-- current 'SlotNo'. Once all 'NumSlots' have passed, the clock is /exhausted/
+-- and most of its methods block indefinitely from then on.
+--
+-- Most notably, 'waitUntilDone' instead blocks until the the clock is
+-- exhausted; so the continuation of that call should expect other threads
+-- using the rest of this clock's methods to be stuck (and eg reap them).
+--
+-- Note: though the wallclock-slot correspondence depends on the ledger state,
+-- we have designed our ledgers so that all nodes necessarily use the same
+-- correspondence in the absence of a Common Prefix violation. So this clock
+-- can indeed know /the/ future.
+data OracularClock m = OracularClock
+    { -- | Returns 'False' if the requested slot is already over
+      --
+      -- Note: blocks indefinitely if clock becomes exhausted before requested
+      -- slot.
+      blockUntilSlot :: SlotNo -> m Bool
+
+      -- | The current delay duration until the onset of the next slot
+      --
+      -- Note: blocks indefinitely if the clock is already exhausted.
+    , delayUntilNextSlot :: m NominalDiffTime
+
+      -- | A mock system time
+      --
+      -- Note: this 'BTime.systemTimeCurrent' blocks indefinitely if clock is
+      -- already exhausted.
+    , finiteSystemTime :: BTime.SystemTime m
+
+      -- | Convert 'Control.Monad.Class.MonadTime.getCurrentTime' to 'SlotNo'
+      --
+      -- INVARIANT: In @io-sim@, the @getCurrentTime@ use (by default) the
+      -- monotonic clock, and it advances /before/ threads correspondingly
+      -- wake-up from a @threadDelay@.
```

Comment:
If you want to use the monotonic clock, perhaps `getMonotonicTime` would be more appropriate?

Reply:
My priority is to use what the implementation uses, which is `defaultSystemTime`, IIUC. And that uses `getCurrentTime`, which is just the wallclock. (Which happens to be monotonic in `io-sim:SimM`.)

I'll add a TODO to update `OracularClock` et al in the test infra to handle clock jumps, so we can eventually test with those. (There's a ticket for this.)

Reply:
I see, fair enough.